conductr_cli/bndl_main.py

```diff
@@ -0,0 +1,597 @@
+                break
+
+        ref_exists = args.tag is not None and os.path.isfile(os.path.join(refs_dir, args.tag))
+
+        if not ref_exists:
+            log.error('bndl: Invalid OCI Image. Cannot find requested tag "{}" in OCI Image'.format(args.tag))
+
+            return 2
+
+        if not has_oci_layout:
+            log.error('bndl: Invalid OCI Image. Missing oci-layout')
+
+            return 2
+
+        bundle_conf = oci_image_bundle_conf(args, component_name)
+        bundle_conf_name = os.path.join(args.name, 'bundle.conf')
+        bundle_conf_data = bundle_conf.encode('UTF-8')
+
+        # bundle.conf must be written first, so we explicitly do that for zip and tar
+
+        if args.use_shazar:
+            with tempfile.NamedTemporaryFile() as zip_file_data:
+                with zipfile.ZipFile(zip_file_data, 'w') as zip_file:
+                    zip_file.writestr(bundle_conf_name, bundle_conf_data)
+                    dir_to_zip(temp_dir, zip_file, args.name)
+
+                zip_file_data.flush()
+                zip_file_data.seek(0)
+                write_with_digest(zip_file_data, output)
+        else:
+            with tarfile.open(fileobj=output, mode='w|') as tar:
+                info = tarfile.TarInfo(name=bundle_conf_name)
+                info.size = len(bundle_conf_data)
+                tar.addfile(tarinfo=info, fileobj=BytesIO(bundle_conf_data))
+
+                for (dir_path, dir_names, file_names) in os.walk(temp_dir):
+                    for file_name in file_names:
+                        path = os.path.join(dir_path, file_name)
+                        name = os.path.join(args.name, os.path.relpath(path, start=temp_dir))
+                        tar.add(path, arcname=name)
+
+        return 0
+    finally:
+        shutil.rmtree(temp_dir)
+
+
+def detect_format_stream(initial_chunk):
+    """
+    Detects the format of a stream given some initial chunk of data. This is fairly crude
+    but does work pretty well with detecting `docker save` streams. OCI detection may need
+    to be expanded as the tooling matures.
+
+    :param initial_chunk:
+    :return: one of 'docker', 'oci-image', or None
+    """
+    try:
+        if b'/oci-layout\x00\x00\x00' in initial_chunk and b'/refs/\x00\x00\x00' in initial_chunk:
+            # tar c on an oci folder (somewhat unreliable)
+            return 'oci-image'
+        elif re.match('^[0-9a-f]{64}[/]', initial_chunk[0:65].decode('UTF-8')):
+            # docker save <image>
+            return 'docker'
+        elif re.match('^[0-9a-f]{64}[.]json$', initial_chunk[0:69].decode('UTF-8')):
+            # docker save <image>:<tag>
+            return 'docker'
+        elif b'/manifest.json\x00\x00\x00' in initial_chunk and b'/layer.tar\x00\x00\x00' in initial_chunk:
+            # tar c on a docker folder (somewhat unreliable)
+            return 'docker'
+        else:
+            return None
+    except UnicodeError:
+        return None
```

Comment:
I think it would be good to have a comment as to what may have caused the `UnicodeError`.